src/PhpAttribute/GenericAnnotationToAttributeConverter.php

```diff
@@ -0,0 +1,108 @@
+use Rector\BetterPhpDocParser\PhpDocInfo\PhpDocInfo;
+use Rector\BetterPhpDocParser\PhpDocInfo\PhpDocInfoFactory;
+use Rector\BetterPhpDocParser\PhpDocManipulator\PhpDocTagRemover;
+use Rector\Naming\Naming\UseImportsResolver;
+use Rector\Php80\NodeFactory\AttrGroupsFactory;
+use Rector\Php80\ValueObject\AnnotationToAttribute;
+use Rector\Php80\ValueObject\DoctrineTagAndAnnotationToAttribute;
+
+/**
+ * @api used in Rector packages
+ */
+final readonly class GenericAnnotationToAttributeConverter
+{
+    public function __construct(
+        private AttrGroupsFactory $attrGroupsFactory,
+        private ReflectionProvider $reflectionProvider,
+        private UseImportsResolver $useImportsResolver,
+        private PhpDocInfoFactory $phpDocInfoFactory,
+        private PhpDocTagRemover $phpDocTagRemover,
+    ) {
+    }
+
+    public function convert(Node $node, AnnotationToAttribute $annotationToAttribute): ?AttributeGroup
+    {
+        if (! $this->isExistingAttributeClass($annotationToAttribute)) {
+            return null;
+        }
+
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNode($node);
+        if (! $phpDocInfo instanceof PhpDocInfo) {
+            return null;
+        }
+
+        $uses = $this->useImportsResolver->resolveBareUses();
+
+        return $this->processDoctrineAnnotationClass($phpDocInfo, $uses, $annotationToAttribute);
+    }
+
+    /**
+     * @param Use_[] $uses
+     */
+    private function processDoctrineAnnotationClass(
+        PhpDocInfo $phpDocInfo,
+        array $uses,
+        AnnotationToAttribute $annotationToAttribute
+    ): ?AttributeGroup {
+        if ($phpDocInfo->getPhpDocNode()->children === []) {
+            return null;
+        }
+
+        $doctrineTagAndAnnotationToAttributes = [];
+        $doctrineTagValueNodes = [];
+
+        foreach ($phpDocInfo->getPhpDocNode()->children as $phpDocChildNode) {
+            if (! $phpDocChildNode instanceof PhpDocTagNode) {
+                continue;
+            }
+
+            if (! $phpDocChildNode->value instanceof DoctrineAnnotationTagValueNode) {
+                continue;
+            }
+
+            $doctrineTagValueNode = $phpDocChildNode->value;
+            if (! $doctrineTagValueNode->hasClassName($annotationToAttribute->getTag())) {
+                continue;
+            }
+
+            $doctrineTagAndAnnotationToAttributes[] = new DoctrineTagAndAnnotationToAttribute(
+                $doctrineTagValueNode,
+                $annotationToAttribute
+            );
+
+            $doctrineTagValueNodes[] = $doctrineTagValueNode;
+        }
+
+        $attributeGroups = $this->attrGroupsFactory->create($doctrineTagAndAnnotationToAttributes, $uses);
+
+        foreach ($doctrineTagValueNodes as $doctrineTagValueNode) {
+            $this->phpDocTagRemover->removeTagValueFromNode($phpDocInfo, $doctrineTagValueNode);
+        }
+
+        return $attributeGroups[0] ?? null;
```

Comment:
this should return array, so use proper array_merge on target since it uses loop that on reuse, it can be multiple

Reply:
Goal is to use single type, to be strict about return value.

Reply:
I create PR for it:

- https://github.com/rectorphp/rector-src/pull/6985

Reply:
Please keep things as they are :pray:

Reply:
Per AttributeGroup -> is singe line, if we are going to only process 1 annotation, then it must be count === 1, that's the way annotation removed === attributeGroup added